homeassistant/components/solaredge/sensor.py

```diff
@@ -79,30 +93,67 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
         _LOGGER.error("Could not retrieve details from SolarEdge API")
         return
 
-    # Create solaredge data service which will retrieve and update the data.
-    data = SolarEdgeData(hass, api, site_id)
+    # Create sensor factory that will create sensors based on sensor_key.
+    sensor_factory = SolarEdgeSensorFactory(platform_name, site_id, api)
 
     # Create a new sensor for each sensor type.
     entities = []
     for sensor_key in config[CONF_MONITORED_CONDITIONS]:
-        sensor = SolarEdgeSensor(platform_name, sensor_key, data)
+        sensor = sensor_factory.create_sensor(sensor_key)
         entities.append(sensor)
 
     add_entities(entities, True)
 
 
+class SolarEdgeSensorFactory:
+    """Factory which creates sensors based on the sensor_key."""
+
+    def __init__(self, platform_name, site_id, api):
+        """Initialize the factory."""
+        self.platform_name = platform_name
+
+        self.overview_data_service = SolarEdgeOverviewDataService(api, site_id)
+        self.details_data_service = SolarEdgeDetailsDataService(api, site_id)
+        self.inventory_data_service = SolarEdgeInventoryDataService(api,
+                                                                    site_id)
+        self.power_flow_data_service = SolarEdgePowerFlowDataService(api,
+                                                                     site_id)
+
+    def create_sensor(self, sensor_key):
+        """Create and return a sensor based on the sensor_key."""
+        sensor = None
+
+        if sensor_key == 'site_details':
```

Comment:
Not sure if I understand correctly. Since each entity class also has its unique data service (depending on the sensor_key) would we not still need an if/elif/else for each sensor_key? I could add a dictionary for the data services in the Factory class as well and use that to retrieve the data_service based on the sensor_key and get the entity class from another dictionary also based on the sensor_key. Would that help here?

Reply:
Ok, so then we can map the sensor key to a tuple of sensor class and data service attribute name. Maybe it's less clear? You can decide what's best.

Reply:
Alright, I hope this was sort or less the idea. I've added a dict with tuples of entity class and data service.